### PR TITLE
Updated the code to throw expeption when main thread is blocked.

### LINKED ISSUE #1681: Workflow | evaluateLibrary not working on android as in the test case.
**Describe the bug**
The evaluateLibrary functions keeps running forever and not returning the results in the android app, however it is working fine in the test case.

**Component**
Core library, SDC library, reference app, or SDC gallery app

**To Reproduce**
Clone the repository [Android-fhir-fork for workflow-eval-library](https://github.com/parthfloyd/android-fhir/tree/workflow-eval-library) // Make sure it's the `workflow-eval-library` branch
1. Run the test case `evaluateImmunityCheck()` in `FhirOperatorLibraryEvaluateJavaTest.kt` 
2. It should be working fine and evaluating correctly.
3. Run the demo app
4. Wait for sync to finish and click on the (+) icon on bottom right (to open add patient questionnaire). 
5. The app crashes after a while after being stuck on the `evaluateLibrary` method in `getQuestionnaireJson()` function of `AddPatientViewModel.kt`.

**Expected behavior**
The method should behave similarly to the test case.

I am not sure if the change is required in the workflow library or some change is required by the implementors in the android app to run this method.

cc: @jingtang10

### PR DESCRIPTION
**Pull request by aditya-07**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1681

**Description**
1. Updated all `runBlocking` calls to `runBlockingOrThrowMainThreadException` that throws `BlockingMainThreadException` when called from main thread.
2. Marked all classes except for `FhirOperator` as internal as `FhirOperator` is supposed to be library's only public interface for developer applications.
3. Added a note regarding the `BlockingMainThreadException` to the functions in `FhirOperator`.
4. Updated tests to be called from worker thread using `runBlockingOnWorkerThread` and added some new tests as well.

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Bug fix

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: workflow/src/main/java/com/google/android/fhir/workflow/FhirEngineRetrieveProvider.kt [hunk 1]
```diff
@@ -56,37 +55,35 @@ class FhirEngineRetrieveProvider(private val fhirEngine: FhirEngine) :
...
+        safeGet(fhirEngine, ResourceType.fromCode(dataType), "urn:oid:$contextValue")
+      )
+    } else if (codePath == "id" && codes != null) {
+      codes.mapNotNull { safeGet(fhirEngine, ResourceType.fromCode(dataType), it.code) }
+    } else {
+      val search = Search(ResourceType.fromCode(dataType))
 
-        // filter by context
-        filterByContext(context, contextPath, contextValue, dataType, search)
+      // filter by context
```
Comment (omarismail94) [inline workflow/src/main/java/com/google/android/fhir/workflow/FhirEngineRetrieveProvider.kt:74]: do these comments add any value here? the function name is pretty descriptive? Here and below.

File: workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+package com.google.android.fhir.workflow
+
+import android.os.Looper
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.runBlocking
+
+/**
+ * Blocks the current thread and run the code in a [CoroutineScope].
```
Comment (omarismail94) [inline workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt:24]: nit: run -> runs

File: workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ * @throws BlockingMainThreadException if the calling thread is main thread.
```
Comment (omarismail94) [inline workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt:25]: nit: calling thread is main -> calling thread is the main

File: workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ */
+internal fun <T> runBlockingOrThrowMainThreadException(block: suspend (CoroutineScope) -> T): T {
+  if (Looper.myLooper() == Looper.getMainLooper()) {
+    throw BlockingMainThreadException()
```
Comment (omarismail94) [inline workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt:29]: Is it possible to add some helper text raised as part of the exception to tell users what they can do to fix the issue?

**Review by omarismail94 — COMMENTED** [review]
Nice change Aditya! Left a couple of comments

File: workflow/src/main/java/com/google/android/fhir/workflow/FhirOperator.kt [hunk 5]
```diff
@@ -185,22 +186,30 @@ class FhirOperator(fhirContext: FhirContext, fhirEngine: FhirEngine) {
...
    */
+  @WorkerThread
   fun evaluateLibrary(libraryUrl: String, expressions: Set<String>): IBaseParameters {
     return evaluateLibrary(libraryUrl, null, null, expressions)
   }
 
   /**
    * The function evaluates a FHIR library against a patient's records.
+   *
+   * NOTE: The api may internally result in a blocking IO operation. The users should call the api
```
Comment (jingtang10) [inline workflow/src/main/java/com/google/android/fhir/workflow/FhirOperator.kt:205]:
```suggestion
   * NOTE: This API may internally result in a blocking IO operation. The user should call the API
```

and below

**Review by jingtang10 — APPROVED** [review]

File: workflow/src/main/java/com/google/android/fhir/workflow/FhirEngineRetrieveProvider.kt (hunk 1 above)
Comment (aditya-07) [reply workflow/src/main/java/com/google/android/fhir/workflow/FhirEngineRetrieveProvider.kt]: Removed.

File: workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt (hunk 2 above)
Comment (aditya-07) [reply workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt]: updated.

File: workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt (hunk 3 above)
Comment (aditya-07) [reply workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt]: updated.

File: workflow/src/main/java/com/google/android/fhir/workflow/FhirOperator.kt (hunk 5 above)
Comment (aditya-07) [reply workflow/src/main/java/com/google/android/fhir/workflow/FhirOperator.kt]: Done for all the APIs.

File: workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt (hunk 4 above)
Comment (aditya-07) [reply workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt:29]: done.

File: workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt [hunk 6]
```diff
@@ -35,4 +35,7 @@ internal fun <T> runBlockingOrThrowMainThreadException(block: suspend (Coroutine
  * The exception that is thrown when an application attempts to perform [runBlocking] operation its
  * main thread.
  */
-class BlockingMainThreadException : RuntimeException()
+internal class BlockingMainThreadException :
+  RuntimeException(
```
Comment (omarismail94) [inline workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt:39]: can you make  `com.google.android.fhir.workflow.FhirOperator` a constant variable and refer to that in the message?

**Review by omarismail94 — APPROVED** [review]
This looks great @aditya-07 ! just one minor nit

File: workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt (hunk 6 above)
Comment (aditya-07) [reply workflow/src/main/java/com/google/android/fhir/workflow/MoreBuilders.kt]: Done. I have also created a constructor for BlockingMainThreadException and message is passed by the caller.

**Review by omarismail94 — APPROVED** [review]

**Pull request merged**